balboa-agent/src/main/scala/com/socrata/balboa/agent/MetricConsumer.scala

```diff
@@ -0,0 +1,280 @@
+    this(directory, metricPublisher, AlphabeticMetricFileProvider(directory.toPath))
+  }
+
+  /**
+    * Attempts to process all the sub directories on the root directories for all possible
+    * metrics.
+    */
+  def run() {
+    logger.info("Looking for metrics files recursively in '{}'", this.directory.getAbsolutePath)
+    val runTimer: Timer.Context = BalboaAgentMetrics.totalRuntime.time
+    val start: Long = System.currentTimeMillis
+    var recordsProcessed: Int = 0
+    // Treat each individual Metric data file as its own isolated run.
+    // We are trying to prevent the failure to process one file from blocking or preventing the processing
+    // of others.
+    fileProvider.provide.foreach { metricsEventLog =>
+      logger.info(s"Processing '${metricsEventLog.getAbsolutePath}'.")
+      var records: util.List[MetricsRecord] = null
+      breakable {
+        try {
+          records = processFile(metricsEventLog)
+        }
+        catch {
+          case e: IOException =>
+            logger.error(s"Error reading records from $metricsEventLog", e)
+            BalboaAgentMetrics.metricsProcessingFailureCounter.inc()
+            val broken: File = new File(metricsEventLog.getAbsolutePath + FileUtils.BROKEN_FILE_EXTENSION)
+            if (!metricsEventLog.renameTo(broken)) {
+              logger.warn(s"Unable to rename broken file $metricsEventLog permissions issue?")
+              BalboaAgentMetrics.renameBrokenFileFailureCounter.inc()
+            }
+            break()
+        }
+        records.asScala.foreach { r =>
+          metricPublisher.create(Fluff(r.entityId), Fluff(r.name), r.value.longValue, r.timestamp, r.metricType)
+        }
+        recordsProcessed += records.size
+        if (!metricsEventLog.delete) {
+          logger.error(s"Unable to delete event log $metricsEventLog - file may be read twice, which is bad.")
+          BalboaAgentMetrics.deleteEventFailureCounter.inc()
+        }
+      }
+    }
+    val processingTime: Long = System.currentTimeMillis - start
+    logger.info(s"Run completed, processed $recordsProcessed in $processingTime ms")
+    BalboaAgentMetrics.metricsEmittedCount.inc(recordsProcessed)
+    BalboaAgentMetrics.metricsEmittedMeter.mark(recordsProcessed)
+    runTimer.stop
+  }
+
+  /**
+    * This method does nothing. As the MetricConsumer does not take unique
+    * ownership of its metricPublisher, it cannot be sure that it is safe to close.
+    */
+  @throws[Exception]
+  def close() {
+  }
+
+  /**
+    * Given a metrics data file, attempt to extract all the metrics from the file and
+    * pushes these metrics into the underlying metricPublisher.
+    *
+    * @param f File to process.
+    * @return A list of { @link MetricsRecord}s that were process.
+    * @throws IOException When there is a problem processing the file.
+    */
+  @throws[IOException]
+  private def processFile(f: File): util.List[MetricsRecord] = {
```

Comment:
Would this be cleaner if we restructured it to just return an iterator instead of building the list in memory?